### PR TITLE
Issue 1330 unsupported echo flags

### PR DESCRIPTION
**Pull request by Gandalf-**
Issue https://github.com/koalaman/shellcheck/issues/1330

Addresses false positives when quoted arguments to echo begin with what
looks like a flag. Now, warn only when the first argument is a
recognized echo flag and when flags are unsupported.

Second attempt, recognizes multiple flags together using regex. New test cases to verify.

## Testing
```
I leaf@willow ~/w/shellcheck (issue_1330_unsupported_echo_flags=)> stack test
...
ShellCheck-0.6.0: Test suite test-shellcheck passed
```

### CODE REVIEW

File: src/ShellCheck/Checks/ShellSupport.hs [hunk 1]
```diff
@@ -238,15 +242,19 @@ checkBashisms = ForShell [Sh, Dash] $ \t -> do
...
                 then
                     when (argString /= "-n") $
                         warnMsg (getId arg) "echo flags besides -n"
                 else
                     warnMsg (getId arg) "echo flags are"
-      where argString = concat $ oversimplify arg
+      where
+          argString = concat $ oversimplify arg
+          flagRegex = mkRegex "^-[eEsn]+$"
+          hasEchoFlags = Just [] == matchRegex flagRegex argString
```
Comment (koalaman) [inline src/ShellCheck/Checks/ShellSupport.hs]: You can simplify this with the function `matches :: String -> Regex -> Bool`, commonly seen with infix notation:  ``argString `matches` matchRegex``

File: src/ShellCheck/Checks/ShellSupport.hs [hunk 2]
```diff
@@ -238,15 +242,19 @@ checkBashisms = ForShell [Sh, Dash] $ \t -> do
         warnMsg id "`<file` to read files is"
 
     bashism t@(T_SimpleCommand _ _ (cmd:arg:_))
-        | t `isCommand` "echo" && "-" `isPrefixOf` argString =
+        | t `isCommand` "echo" && hasEchoFlags =
             unless ("--" `isPrefixOf` argString) $ -- echo "-----"
```
Comment (koalaman) [inline src/ShellCheck/Checks/ShellSupport.hs]: I guess this part is redundant now

File: src/ShellCheck/Checks/ShellSupport.hs [hunk 3]
```diff
@@ -238,15 +242,19 @@ checkBashisms = ForShell [Sh, Dash] $ \t -> do
...
                 if isDash
                 then
                     when (argString /= "-n") $
                         warnMsg (getId arg) "echo flags besides -n"
                 else
                     warnMsg (getId arg) "echo flags are"
-      where argString = concat $ oversimplify arg
+      where
+          argString = concat $ oversimplify arg
```
Comment (koalaman) [inline src/ShellCheck/Checks/ShellSupport.hs]: You could optionally replace `concat $ oversimplify arg` with the less heavy handed version `getLiteralString arg`. This returns a `Maybe String`, with Nothing if the value is not purely literal, such as if it contains a variable (`oversimplify` instead arbitrarily inserts `"${VAR}"` in place of the variable)

Alternatively, just leave it.

**Comment by Gandalf-** [comment]
Thanks for the feedback! I'll update the commit this evening.

**Comment by Gandalf-** [comment]
# Testing
```
I leaf@willow ~/w/shellcheck (issue_1330_unsupported_echo_flags=)> stack test
...

ShellCheck-0.6.0: Test suite test-shellcheck passed
```

**Comment by koalaman** [comment]
Nice! This got rid of a bunch of false positives.

**Pull request merged**